webtau-core/src/main/java/com/twosigma/webtau/expectation/equality/handlers/DateAndStringCompareToHandler.java

```diff
@@ -0,0 +1,168 @@
+        private final ActualPath actualPath;
+        private final TemporalAccessor actual;
+        private final Object expected;
+        private final boolean isEqualOnly;
+
+        Comparator(CompareToComparator compareToComparator, ActualPath actualPath,
+                   Object actual, Object expected, boolean isEqualOnly) {
+            this.compareToComparator = compareToComparator;
+            this.actualPath = actualPath;
+            this.actual = actualToTemporalAccessor(actual);
+            this.expected = expected;
+            this.isEqualOnly = isEqualOnly;
+        }
+
+        void compare() {
+            if (actual instanceof LocalDate && expected instanceof LocalDate) {
+                compareLocalDates((LocalDate) actual, (LocalDate) expected);
+            } else if (actual instanceof ZonedDateTime && expected instanceof LocalDate) {
+                compareZonedDateTimeAndLocalDate((ZonedDateTime) actual, (LocalDate) expected);
+            } else if (actual instanceof ZonedDateTime && expected instanceof ZonedDateTime) {
+                compareZonedDateTimes((ZonedDateTime) actual, (ZonedDateTime) expected);
+            } else {
+                throw new UnsupportedOperationException("combination is not supported:\n" +
+                        renderActualExpected(actual, expected));
+            }
+        }
+
+        private void compareZonedDateTimes(ZonedDateTime actual, ZonedDateTime expected) {
+            report(actual.compareTo(expected), renderActualExpectedNormalized(actual, expected));
+        }
+
+        private void compareZonedDateTimeAndLocalDate(ZonedDateTime actual, LocalDate expected) {
+            report(actual.toLocalDate().compareTo(expected), renderActualExpected(actual, expected));
+        }
+
+        private void compareLocalDates(LocalDate actual, LocalDate expected) {
+            report(actual.compareTo(expected), renderActualExpected(actual, expected));
+        }
+
+        private void report(int compareTo, String message) {
+            if (isEqualOnly) {
+                compareToComparator.reportEqualOrNotEqual(DateAndStringCompareToHandler.this,
+                        compareTo == 0, actualPath, message);
+            } else {
+                compareToComparator.reportCompareToValue(DateAndStringCompareToHandler.this,
+                        compareTo, actualPath, message);
+            }
+        }
+
+        private TemporalAccessor actualToTemporalAccessor(Object actual) {
+            String actualAsText = (String) actual;
+
+            for (FormatParser parser: parsers) {
+                try {
+                    return parser.convert(actualAsText);
+                } catch (DateTimeParseException ignored) {
+                }
+            }
+
+            throw new UnsupportedOperationException("cannot parse " + actualAsText + "\navailable formats:\n" +
+                    parsers.stream().map(FormatParser::toString).collect(Collectors.joining("\n")));
```

Comment:
Instead of all this, could do `String.join('\n', parsers)`

Reply:
keep forgetting about this one! thanks.

Reply:
Hmm, String.join expects `Iterable<? extends CharSequence>`. `parsers` is `List<FormatParser>`

Reply:
Oh, yeah, this always annoyed me and I always forget.  On many occasions, I've ended up writing my own version of `String.join` which takes essentially an Iterable of objects.  At least where I couldn't use Eclipse Collections (or Kotlin) which handle this problem natively.